gobblin-utility/src/main/java/gobblin/util/request_allocation/AllocatedRequestsBase.java

```diff
@@ -0,0 +1,75 @@
+package gobblin.util.request_allocation;
+
+import java.util.Iterator;
+import java.util.Random;
+
+import com.google.common.base.Function;
+
+import lombok.Getter;
+
+
+/**
+ * A basic implementation of {@link AllocatedRequests}.
+ * @param <T>
+ */
+public class AllocatedRequestsBase<T extends Request<T>> implements AllocatedRequests<T> {
+
+  private final Iterator<TAndRequirement<T>> underlying;
+  private final double[] currentRequirement;
+
+  public AllocatedRequestsBase(Iterator<TAndRequirement<T>> underlying, ResourcePool resourcePool) {
+    this.underlying = underlying;
+    this.currentRequirement = new double[resourcePool.getNumDimensions()];
+  }
+
+  @Override
+  public ResourceRequirement totalResourcesUsed() {
+    return new ResourceRequirement(this.currentRequirement);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return this.underlying.hasNext();
+  }
+
+  @Override
+  public T next() {
+    TAndRequirement<T> nextElement = this.underlying.next();
+    VectorAlgebra.addVector(this.currentRequirement, nextElement.getResourceRequirement().getResourceVector(), 1.0,
+        this.currentRequirement);
+    return nextElement.getT();
+  }
+
+  @Override
+  public void remove() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Stores and element and its {@link ResourceRequirement}.
+   */
+  @Getter
+  public static class TAndRequirement<T> {
+    public static final Random RANDOM = new Random();
+
+    private final T t;
+    private final ResourceRequirement resourceRequirement;
+    private final long id;
+
+    TAndRequirement(T t, ResourceRequirement resourceRequirement) {
+      this.t = t;
+      this.resourceRequirement = resourceRequirement;
+      this.id = RANDOM.nextLong();
```

Comment:
This is used in `ConcurrentBoundedPriorityIterable` to make requests that are equal according to the prioritizer still different according to the internal `TreeMap`. The probability of collision is ridiculously small (long is 64 bits, even with 1million requests, the probability of collision is 2.7e-8, obviously assuming the RNG is good enough).